addons/pkg/util/watch_remote.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package util
+
+import (
+	"context"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/cluster-api/controllers/remote"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// RemoteObjectTracker is a helper struct to deal with watching remote clusters
+type RemoteObjectTracker struct {
```

Comment:
question: wondering if we could just reuse `*remote.ClusterCacheTracker` ?

Reply:
I think that ```*remote.ClusterCacheTracker``` is written for cluster-api specific purpose and re-using it is not a good idea. For example, it creates a cache for the cluster, starts health checked for teh cluster, etc that we don't need for our purposes: 
https://github.com/kubernetes-sigs/cluster-api/blob/8941f168e9b45c0d94267cd957eaa3d1bba36a97/controllers/remote/cluster_cache.go#L176

https://github.com/kubernetes-sigs/cluster-api/blob/8941f168e9b45c0d94267cd957eaa3d1bba36a97/controllers/remote/cluster_cache.go#L201

https://github.com/kubernetes-sigs/cluster-api/blob/8941f168e9b45c0d94267cd957eaa3d1bba36a97/controllers/remote/cluster_cache.go#L207

Reply:
Cache knows how to load Kubernetes objects, fetch informers to request to receive events for Kubernetes objects (at a low-level)
```
type Cache interface {
	// Cache acts as a client to objects stored in the cache.
	client.Reader

	// Cache loads informers and adds field indices.
	Informers
}
```

If just using what you have now, I am not sure how informer knows what apiserver endpoint to watch the events.
```
if err := input.Watcher.Watch(&source.Kind{Type: input.Kind}, input.EventHandler, input.Predicates...); err != nil {
		return errors.Wrap(err, "error creating watch")
	}
```

Cluster api uses the following:
```
	if err := input.Watcher.Watch(source.NewKindWithCache(input.Kind, a.cache), input.EventHandler, input.Predicates...); err != nil {
		return errors.Wrap(err, "error creating watch")
	}
```
`a.cache` actually holds the necessary information tells informer how to access a remote cluster apiserver.

Reply:
I’m using manager’s cache.
When the controller is added to the manager, the manager’s cache is injected into the controller. 

Per the comment in Source interface [here](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/source/source.go#L54), “Users may build their own Source implementations.  If their implementations implement any of the inject package interfaces, the dependencies will be injected by the Controller when Watch is called.”

NewKindWithCache used by cluster-api ([here](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/source/source.go#L72)) creates a Source without InjectCache.

If not, the Injected cache by controller-runtime ([here](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/source/source.go#L189)) will be used which is what I've been doing.

So logically my implementation works (and I have tested it with multiple clusters). 

My only concern now is that whether using the manager’s cache for all remote clusters scale well, or better to set up a new cache per cluster as cluster-api does ([here](https://github.com/kubernetes-sigs/cluster-api/blob/main/controllers/remote/cluster_cache.go#L181))

Reply:
Update on this: I asked in the public channel and I'll update the code to use separate cache per cluster.